gateway/src/node/client_handling/websocket/connection_handler/ecash/credential_sender.rs

```diff
@@ -0,0 +1,216 @@
+            credential,
+            address,
+            api_clients,
+            proposal_id: None,
+        }
+    }
+}
+
+pub(crate) struct CredentialSender<St: Storage> {
+    cred_receiver: UnboundedReceiver<PendingCredential>,
+    storage: St,
+    nyxd_client: Arc<RwLock<DirectSigningHttpRpcNyxdClient>>,
+}
+
+impl<St> CredentialSender<St>
+where
+    St: Storage + 'static,
+{
+    pub(crate) fn new(
+        cred_receiver: UnboundedReceiver<PendingCredential>,
+        storage: St,
+        nyxd_client: Arc<RwLock<DirectSigningHttpRpcNyxdClient>>,
+    ) -> Self {
+        CredentialSender {
+            cred_receiver,
+            storage,
+            nyxd_client,
+        }
+    }
+
+    async fn create_proposal(
+        &self,
+        pending: &mut PendingCredential,
+    ) -> Result<(), RequestHandlingError> {
+        if pending.proposal_id.is_some() {
+            log::trace!("Proposal already created");
+            return Ok(());
+        }
+        if pending.credential.data.typ.is_free_pass() {
+            //no proposal for freepasses
+            return Ok(());
+        }
+        let res = self
+            .nyxd_client
+            .write()
+            .await
+            .prepare_credential(
+                pending.credential.data.serial_number_b58(),
+                pending.address.to_string(),
+                None,
+            )
+            .await?;
+        let proposal_id = find_attribute(&res.logs, "wasm", BANDWIDTH_PROPOSAL_ID)
+            .ok_or(RequestHandlingError::ProposalIdError {
+                reason: String::from("proposal id not found"),
+            })?
+            .value
+            .parse::<u64>()
+            .map_err(|_| RequestHandlingError::ProposalIdError {
+                reason: String::from("proposal id could not be parsed to u64"),
+            })?;
+
+        let proposal = self
+            .nyxd_client
+            .read()
+            .await
+            .query_proposal(proposal_id)
+            .await?;
+        if !pending
+            .credential
+            .matches_serial_number(&proposal.description)?
+        {
+            return Err(RequestHandlingError::ProposalIdError {
+                reason: String::from("proposal has different serial number"),
+            });
+        }
+        pending.proposal_id = Some(proposal_id);
+        Ok(())
+    }
+
+    async fn send_credential(pending: &mut PendingCredential) -> Result<(), RequestHandlingError> {
+        if pending.credential.data.typ.is_ticketbook() && pending.proposal_id.is_none() {
+            return Err(RequestHandlingError::ProposalIdError {
+                reason: "proposal id is absent".to_string(),
+            });
+        }
+        let request = VerifyEcashCredentialBody::new(
+            pending.credential.data.clone(),
+            pending.address.clone(),
+            pending.proposal_id,
+        );
+        let mut failed_api = Vec::new();
+        for client in &pending.api_clients {
```

Comment:
try to parallelise it like i explained in the ecash/mod.rs file

Reply:
TODO after merge?